src/trace/dma-trace.c

```diff
@@ -126,17 +126,52 @@ static enum task_state trace_work(void *data)
  */
 int dma_trace_init_early(struct sof *sof)
 {
+	int ret;
+
+	/* If this assert is wrong then traces have been corrupting
+	 * random parts of memory. Some functions run before _and_ after
+	 * DMA trace initialization and we don't want to ask them to
+	 * never trace. So dma_trace_initialized() must be either
+	 * clearly false/NULL or clearly true, we can't tolerate random
+	 * uninitialized values in sof->dmat etc.
+	 */
+	assert(!dma_trace_initialized(sof->dmat));
+
 	sof->dmat = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*sof->dmat));
+	if (!sof->dmat) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
 	dma_sg_init(&sof->dmat->config.elem_array);
 	spinlock_init(&sof->dmat->lock);
 
 	ipc_build_trace_posn(&sof->dmat->posn);
 	sof->dmat->msg = ipc_msg_init(sof->dmat->posn.rhdr.hdr.cmd,
 				      sizeof(sof->dmat->posn));
-	if (!sof->dmat->msg)
-		return -ENOMEM;
+	if (!sof->dmat->msg) {
+		ret = -ENOMEM;
+		goto err;
+	}
 
 	return 0;
+
+err:
+	mtrace_printf(LOG_LEVEL_ERROR,
+		      "dma_trace_init_early() failed: %d", ret);
+
+	/* This crashes for some unknown reason. Circular initialization
+	 * dependencies? Not freeing this means wasting a small amount
+	 * of memory in a situation with a much "bigger problem":
+	 * tracing failed to initialize! The longer term direction is
+	 * Zephyr, so this allocator is deprecated.
+	 */
+	if (false)
+		rfree(sof->dmat);
```

Comment:
it crashes because `SOF_MEM_ZONE_SYS_SHARED` and `SOF_MEM_ZONE_SYS` zones aren't supposed to be freed.